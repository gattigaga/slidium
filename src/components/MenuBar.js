import React, { Component, Children, cloneElement } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Container = styled.div`
  box-sizing: border-box;
  width: 100%;
  display: flex;
  margin-top: 4px;
`;

class MenuBar extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedMenu: -1
    };

    this.toggleMenu = this.toggleMenu.bind(this);
  }

  /**
   * Toggle menu open or close
   *
   * @param {number} index - Menu index
   * @memberof MenuBar
   */
  toggleMenu(index) {
    this.setState(prevState => {
      const { selectedMenu } = prevState;
      // Check if user was click to open it or not
      const isOpen = selectedMenu === index;

      return {
        selectedMenu: isOpen ? -1 : index
      };
    });
  }

  render() {
    const { selectedMenu } = this.state;
    const { children } = this.props;

    return (
      <Container>
        {Children.map(children, (child, index) => {
          return cloneElement(child, {
            isOpen: selectedMenu === index,
            onClick: () => this.toggleMenu(index)
          });
        })}
      </Container>
    );
  }
}

MenuBar.propTypes = {
  children: PropTypes.node
};

export default MenuBar;
