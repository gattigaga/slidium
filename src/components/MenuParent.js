import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Container = styled.div`
  width: auto;
  padding: 4px 8px;
  cursor: default;
  position: relative;
  user-select: none;

  &:hover {
    background: #eee;
  }
`;

const Label = styled.span`
  font-size: 12px;
  font-family: Roboto;
  color: black;
`;

const LabelActive = Container.extend`
  position: absolute;
  top: -28px;
  left: -1px;
  z-index: 3;
  background: white;
  border: 1px solid #ccc;
  border-bottom: 0px;
  padding-bottom: 5px;

  &:hover {
    background: white;
  }
`;

const Tray = styled.span`
  position: absolute;
  min-width: 192px;
  top: 100%;
  left: 0%;
  background: white;
  border: 1px solid #ccc;
  padding: 8px 0px;
  box-shadow: 1px 1px 8px rgba(0, 0, 0, 0.2);
  z-index: 2;
`;

const MenuParent = ({ label, children, isOpen, onClick }) => (
  <Container onClick={onClick}>
    <Label>{label}</Label>
    {isOpen && (
      <Tray>
        <LabelActive>
          <Label>{label}</Label>
        </LabelActive>
        {children}
      </Tray>
    )}
  </Container>
);

MenuParent.propTypes = {
  label: PropTypes.string,
  children: PropTypes.node,
  onClick: PropTypes.func,
  isOpen: PropTypes.bool
};

MenuParent.defaultProps = {
  label: "Menu"
};

export default MenuParent;
