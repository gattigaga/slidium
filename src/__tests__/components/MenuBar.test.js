import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import MenuBar from "components/MenuBar";
import MenuParent from "components/MenuParent";

let wrapper;

beforeEach(() => {
  wrapper = shallow(
    <MenuBar>
      <MenuParent label="New" />
      <MenuParent label="Edit" />
      <MenuParent label="View" />
    </MenuBar>
  );
});

describe("MenuBar", () => {
  it("renders empty child", () => {
    wrapper = shallow(<MenuBar />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders with child", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("toggle first menu", () => {
    // Open Menu
    wrapper
      .find("MenuParent")
      .first()
      .simulate("click");
    expect(toJSON(wrapper)).toMatchSnapshot();

    // Close Menu
    wrapper
      .find("MenuParent")
      .first()
      .simulate("click");
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("open first menu, autoclose it, and then open second menu", () => {
    // Open Menu
    wrapper
      .find("MenuParent")
      .first()
      .simulate("click");
    expect(toJSON(wrapper)).toMatchSnapshot();

    // Close Menu
    wrapper
      .find("MenuParent")
      .at(1)
      .simulate("click");
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
