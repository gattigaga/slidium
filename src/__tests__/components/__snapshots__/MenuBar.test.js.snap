// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuBar open first menu, autoclose it, and then open second menu 1`] = `
<styled.div>
  <MenuParent
    isOpen={true}
    key=".0"
    label="New"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".1"
    label="Edit"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".2"
    label="View"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`MenuBar open first menu, autoclose it, and then open second menu 2`] = `
<styled.div>
  <MenuParent
    isOpen={false}
    key=".0"
    label="New"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={true}
    key=".1"
    label="Edit"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".2"
    label="View"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`MenuBar renders empty child 1`] = `<styled.div />`;

exports[`MenuBar renders with child 1`] = `
<styled.div>
  <MenuParent
    isOpen={false}
    key=".0"
    label="New"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".1"
    label="Edit"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".2"
    label="View"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`MenuBar toggle first menu 1`] = `
<styled.div>
  <MenuParent
    isOpen={true}
    key=".0"
    label="New"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".1"
    label="Edit"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".2"
    label="View"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`MenuBar toggle first menu 2`] = `
<styled.div>
  <MenuParent
    isOpen={false}
    key=".0"
    label="New"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".1"
    label="Edit"
    onClick={[Function]}
  />
  <MenuParent
    isOpen={false}
    key=".2"
    label="View"
    onClick={[Function]}
  />
</styled.div>
`;
